R/epiparameter.R

```diff
@@ -0,0 +1,996 @@
+#' Constructor for `<epiparameter>` class
+#'
+#' @description Create an `<epiparameter>` object. The
+#' constructor will search whether parameters of the probability distribution
+#' are supplied and if not look to see whether they can be inferred/extracted/
+#' converted from summary statistics provided. It will also convert the
+#' probability distribution (`prob_dist`) and its parameters
+#' (`prob_dist_params`) into an S3 class, either a `distribution` object from
+#' `{distributional}` when `discretise = FALSE`, or a `distcrete` object from
+#' `{distcrete}` when `discretise = TRUE`.
+#'
+#' @inheritParams epiparameter
+#'
+#' @inherit epiparameter return
+#' @keywords internal
+new_epiparameter <- function(disease = character(),
+                             pathogen = character(),
+                             epi_name = character(),
+                             prob_distribution = list(),
+                             uncertainty = list(),
+                             summary_stats = list(),
+                             citation = character(),
+                             metadata = list(),
+                             method_assess = list(),
+                             notes = character(),
+                             auto_calc_params = logical(),
+                             ...) {
+  if (auto_calc_params && is.character(prob_distribution)) {
+    # calculate parameters if not provided
+    prob_distribution_params <- .calc_dist_params(
+      prob_distribution = prob_distribution,
+      summary_stats = summary_stats,
+      sample_size = metadata$sample_size
+    )
+    if (!anyNA(prob_distribution_params)) {
+      prob_distribution <- create_prob_distribution(
+        prob_distribution = prob_distribution,
+        prob_distribution_params = prob_distribution_params,
+        ...
+      )
+      message(
+        "Parameterising the probability distribution with the summary ",
+        "statistics.\n Probability distribution is assumed not to be ",
+        "discretised or truncated."
+      )
+    }
+    if (!inherits(prob_distribution, c("distribution", "distcrete"))) {
+      message("Unparameterised <epiparameter> object")
+    }
+  }
+
+  if (epi_name == "offspring_distribution") {
```

Comment:
I have made some improvements to this in PR #401 to try and match non-delay distributions, which currently include offspring distributions and case fatality risks. 

The data dictionary used in the JSON validation workflow uses an `enum` field to check that the `epi_name` is within a predefined set https://github.com/epiverse-trace/epiparameter/blob/main/inst/extdata/data_dictionary.json#L23.

However, this is a temporary solution and needs to be improved to be more scalable as more non-delay distributions are added to the package, and to accommodate users creating `<epiparameter>` objects with a variety of parameter names. We could impose that the `epi_name` argument must also be within the same set as defined in the data dictionary in when specified by the user in `epiparameter()` but I don't think there would be that much benefit, while it could hinder users and put more development burden adding more parameter types. 

Happy to discuss this further to find a more optimal solution. This can optionally be filed as an issue to continue discussion after the package review.